tests/test_products.py

```diff
@@ -0,0 +1,989 @@
+        arr = np.ma.array(np.zeros(shape, dtype=np.float32))
+        output_arr = NDVI()._force_nodata_where_it_was_in_original_raster(arr, raster, raster.band_names)
+        self.assertTrue((output_arr == 0).all())
+        self.assertFalse(output_arr.mask.any())
+
+    def test_setting_minimal_value_for_nodata_uint16(self):
+        raster = make_test_raster(42, ['nir', 'red'], dtype=np.uint16)
+        array = raster.image.data
+        mask = raster.image.mask
+        mask[:, :, :] = True
+        raster = raster.copy_with(image=np.ma.array(data=array, mask=mask))
+        shape = (1, raster.height, raster.width)
+        arr = np.ma.array(np.zeros(shape))
+        output_arr = NDVI()._force_nodata_where_it_was_in_original_raster(arr, raster, raster.band_names)
+        self.assertTrue(output_arr.mask.all())
+
+    def test_setting_minimal_value_for_nodata_float32(self):
+        raster = make_test_raster(0.42, ['nir', 'red'], dtype=np.float32)
+        array = raster.image.data
+        mask = raster.image.mask
+        mask[:, :, :] = True
+        raster = raster.copy_with(image=np.ma.array(data=array, mask=mask))
+        shape = (1, raster.height, raster.width)
+        arr = np.ma.array(np.zeros(shape))
+        output_arr = NDVI()._force_nodata_where_it_was_in_original_raster(arr, raster, raster.band_names)
+        self.assertTrue(output_arr.mask.all())
+
+
+class TestBandsMatching(unittest.TestCase):
+
+    def test_it_matches_all_bands(self):
+        self.assertTrue(set(ProductsFactory.get_matchings(['blue', 'nir', 'red'],
+                                                          sensor_bands_info())).issuperset(['NDVI', 'EVI2']))
+        self.assertTrue(set(ProductsFactory.get_matchings(['nir', 'red'],
+                                                          sensor_bands_info())).issuperset(['NDVI', 'EVI2']))
+        self.assertNotIn('EXR', ProductsFactory.get_matchings(['nir', 'red'], sensor_bands_info()))
+        self.assertEqual(set(ProductsFactory.get_matchings(['red'], sensor_bands_info())), {'SingleBand'})
+
+
+class TestNDVIStraite(unittest.TestCase):
+
+    def test_ndvi(self):
+        raster = NDVI().apply(sensor_bands_info(), multi_raster_8b())
+        self.assertEqual(raster.num_bands, 1)
+        self.assertEqual(raster.band_names, NDVI.output_bands)
+        self.assertEqual(raster.dtype, NDVI.type)
+        self.assertEqual(raster.height, multi_raster_8b().height)
+        self.assertEqual(raster.width, multi_raster_8b().width)
+        expected_value = (multi_values_8b['nir'] - multi_values_8b['red']) / (
+            multi_values_8b['nir'] + multi_values_8b['red']
+        )
+        self.assertTrue((raster.image.data == expected_value).all())
+
+    def test_NDVI_product_base(self):
+        product_generator = NDVI()
+        product = product_generator.apply(sensor_bands_info(), multi_raster_8b(), metadata=True)
+        self.assertEqual(product.name, product_generator.name)
```

Comment:
This block of assert is repeated for different products. Does it make sense to test it in all cases, or perhaps it would be enough to test the parent class?

Reply:
the best could be using `pytest parametrize`